File: src/main/java/com/paragon/Paragon.java
package com.paragon;

import com.paragon.api.event.EventParser;
import com.paragon.client.managers.*;
import com.paragon.client.managers.alt.AltManager;
import com.paragon.client.managers.rotation.RotationManager;
import com.paragon.client.managers.social.SocialManager;
import com.paragon.client.systems.ui.console.Console;
import com.paragon.client.systems.ui.panel.PanelGUI;
import com.paragon.client.systems.ui.taskbar.Taskbar;
import com.paragon.client.systems.ui.window.WindowGUI;
import me.wolfsurge.cerauno.EventBus;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod(name = Paragon.modName, modid = Paragon.modID, version = Paragon.modVersion)
public class Paragon {

    public static final String modName = "Paragon";
    public static final String modID = "paragon";
    public static final String modVersion = "1.0.0";

    @Mod.Instance
    public static Paragon INSTANCE = new Paragon();

    // Client stuff
    private Logger logger;
    private final EventBus eventBus = new EventBus();

    // Managers
    private FontManager fontManager;
    private StorageManager storageManager;
    private ModuleManager moduleManager;
    private CommandManager commandManager;
    private EventParser eventParser;
    private PopManager popManager;
    private SocialManager socialManager;
    private AltManager altManager;
    private RotationManager rotationManager;

    // GUIs
    private Taskbar taskbar;
    private PanelGUI panelGUI;
    private WindowGUI windowGUI;
    private Console console;

    @Mod.EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        logger = LogManager.getLogger("Paragon");

        fontManager = new FontManager();
    }

    @Mod.EventHandler
    public void init(FMLInitializationEvent event) {
        getLogger().info("Starting Paragon " + modVersion + " initialisation");

        // Set up managers
        storageManager = new StorageManager();

        moduleManager = new ModuleManager();
        commandManager = new CommandManager();

        eventParser = new EventParser();
        popManager = new PopManager();

        socialManager = new SocialManager();
        getStorageManager().loadSocial();

        altManager = new AltManager();
        getStorageManager().loadAlts();

        rotationManager = new RotationManager();

        // Set up GUIs and elements
        taskbar = new Taskbar();
        panelGUI = new PanelGUI();
        windowGUI = new WindowGUI();
        console = new Console("Paragon Console", 400, 300);

        getLogger().info("Paragon Initialised Successfully");
    }

    /**
     * Gets the logger
     * @return The logger
     */
    public Logger getLogger() {
        return logger;
    }

    /**
     * Gets the event bus
     * @return The event bus
     */
    public EventBus getEventBus() {
        return eventBus;
    }

    /**
     * Gets the font manager
     * @return The font manager
     */
    public FontManager getFontManager() {
        return fontManager;
    }

    /**
     * Gets the storage manager
     * @return The storage manager
     */
    public StorageManager getStorageManager() {
        return storageManager;
    }

    /**
     * Gets the module manager
     * @return The client's module manager
     */
    public ModuleManager getModuleManager() {
        return moduleManager;
    }

    /**
     * Gets the command manager
     * @return The command manager
     */
    public CommandManager getCommandManager() {
        return commandManager;
    }

    /**
     * Gets the event parser
     * @return The event parser
     */
    public EventParser getEventParser() {
        return eventParser;
    }

    /**
     * Gets the pop manager
     * @return The pop manager
     */
    public PopManager getPopManager() {
        return popManager;
    }

    /**
     * Gets the social manager
     * @return The social manager
     */
    public SocialManager getSocialManager() {
        return socialManager;
    }

    /**
     * Gets the alt manager
     * @return The alt manager
     */
    public AltManager getAltManager() {
        return altManager;
    }

    /**
     * Gets the rotation manager
     * @return The rotation manager
     */
    public RotationManager getRotationManager() {
        return rotationManager;
    }

    /**
     * Gets the taskbar
     * @return The taskbar
     */
    public Taskbar getTaskbar() {
        return taskbar;
    }

    /**
     * Gets the panel GUI
     * @return The panel GUI
     */
    public PanelGUI getPanelGUI() {
        return panelGUI;
    }

    /**
     * Gets the window GUI
     * @return The window GUI
     */
    public WindowGUI getWindowGUI() {
        return windowGUI;
    }

    /**
     * Gets the console
     * @return The console
     */
    public Console getConsole() {
        return console;
    }
}

File: src/main/kotlin/com/paragon/Paragon.kt
package com.paragon

import com.paragon.bus.EventBus
import com.paragon.impl.event.EventFactory
import com.paragon.impl.managers.*
import com.paragon.impl.ui.configuration.ConfigurationGUI
import com.paragon.impl.ui.configuration.GuiImplementation
import com.paragon.impl.ui.configuration.retrowindows.Windows98
import com.paragon.impl.ui.console.Console
import com.paragon.impl.ui.taskbar.Taskbar
import com.paragon.util.render.font.FontUtil
import net.minecraft.client.Minecraft
import net.minecraftforge.common.ForgeVersion
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.lwjgl.opengl.Display
import java.awt.Desktop
import java.net.URI
import javax.swing.JOptionPane

@Mod(name = com.paragon.Paragon.Companion.modName, modid = com.paragon.Paragon.Companion.modID, version = com.paragon.Paragon.Companion.modVersion)
class Paragon {

    @EventHandler
    fun preInit(event: FMLPreInitializationEvent?) {
        if (ForgeVersion.buildVersion < 2860) {
            JOptionPane.showMessageDialog(
                null, "Forge version is too old. Paragon requires Forge to be at least build 2860.", "Outdated Forge!", JOptionPane.ERROR_MESSAGE
            )

            Desktop.getDesktop().browse(
                URI("https://files.minecraftforge.net/net/minecraftforge/forge/index_1.12.2.html")
            )

            Minecraft.getMinecraft().shutdown()

            // When trying to exit throws an exception lmao
            Display.destroy()

            return
        }

        eventParser = EventFactory()
        FontUtil.init()
    }

    @EventHandler
    fun init(event: FMLInitializationEvent?) {
        logger.info("Starting Paragon ${modVersion} initialisation")

        //  ________  ________  ________  ________  ________  ________  ________            _____      ________      ________
        // |\   __  \|\   __  \|\   __  \|\   __  \|\   ____\|\   __  \|\   ___  \         / __  \    |\   __  \    |\   __  \
        // \ \  \|\  \ \  \|\  \ \  \|\  \ \  \|\  \ \  \___|\ \  \|\  \ \  \\ \  \       |\/_|\  \   \ \  \|\  \   \ \  \|\  \
        //  \ \   ____\ \   __  \ \   _  _\ \   __  \ \  \  __\ \  \\\  \ \  \\ \  \      \|/ \ \  \   \ \  \\\  \   \ \  \\\  \
        //   \ \  \___|\ \  \ \  \ \  \\  \\ \  \ \  \ \  \|\  \ \  \\\  \ \  \\ \  \          \ \  \ __\ \  \\\  \ __\ \  \\\  \
        //    \ \__\    \ \__\ \__\ \__\\ _\\ \__\ \__\ \_______\ \_______\ \__\\ \__\          \ \__\\__\ \_______\\__\ \_______\
        //     \|__|     \|__|\|__|\|__|\|__|\|__|\|__|\|_______|\|_______|\|__| \|__|           \|__\|__|\|_______\|__|\|_______|

        println(
            "\n ________  ________  ________  ________  ________  ________  ________            _____      ________      ________     \n" + "|\\   __  \\|\\   __  \\|\\   __  \\|\\   __  \\|\\   ____\\|\\   __  \\|\\   ___  \\         / __  \\    |\\   __  \\    |\\   __  \\    \n" + "\\ \\  \\|\\  \\ \\  \\|\\  \\ \\  \\|\\  \\ \\  \\|\\  \\ \\  \\___|\\ \\  \\|\\  \\ \\  \\\\ \\  \\       |\\/_|\\  \\   \\ \\  \\|\\  \\   \\ \\  \\|\\  \\   \n" + " \\ \\   ____\\ \\   __  \\ \\   _  _\\ \\   __  \\ \\  \\  __\\ \\  \\\\\\  \\ \\  \\\\ \\  \\      \\|/ \\ \\  \\   \\ \\  \\\\\\  \\   \\ \\  \\\\\\  \\  \n" + "  \\ \\  \\___|\\ \\  \\ \\  \\ \\  \\\\  \\\\ \\  \\ \\  \\ \\  \\|\\  \\ \\  \\\\\\  \\ \\  \\\\ \\  \\          \\ \\  \\ __\\ \\  \\\\\\  \\ __\\ \\  \\\\\\  \\ \n" + "   \\ \\__\\    \\ \\__\\ \\__\\ \\__\\\\ _\\\\ \\__\\ \\__\\ \\_______\\ \\_______\\ \\__\\\\ \\__\\          \\ \\__\\\\__\\ \\_______\\\\__\\ \\_______\\\n" + "    \\|__|     \\|__|\\|__|\\|__|\\|__|\\|__|\\|__|\\|_______|\\|_______|\\|__| \\|__|           \\|__\\|__|\\|_______\\|__|\\|_______|"
        )

        storageManager = StorageManager()
        logger.info("StorageManager initialised")

        // Module /  Commands

        moduleManager = ModuleManager()
        logger.info("ModuleManager initialised")

        commandManager = CommandManager()
        logger.info("CommandManager initialised")

        pluginManager = PluginManager()
        logger.info("PluginManager initialised")

        pluginManager.onLoad()
        logger.info("Plugins loaded")

        // Misc client stuff

        altManager = AltManager()
        logger.info("AltManager initialised")

        capeManager = CapeManager()

        notificationManager = NotificationManager()
        logger.info("NotificationManager initialised")

        friendManager = FriendManager()
        logger.info("SocialManager initialised")

        // Event / Ingame stuff

        popManager = PopManager()
        logger.info("PopManager initialised")

        rotationManager = RotationManager()
        logger.info("RotationManager initialised")

        // GUIs

        taskbar = Taskbar
        logger.info("Taskbar Initialised")

        windows98GUI = Windows98()
        logger.info("Windows98 GUI Initialised")

        console = Console("Paragon Console", 400f, 300f)
        logger.info("Console Initialised")

        configurationGUI = ConfigurationGUI()
        logger.info("Configuration GUI Initialised")

        // Load

        storageManager.loadModules("current")
        logger.info("Modules Loaded")

        storageManager.loadSocial()
        logger.info("Social Loaded")

        storageManager.loadAlts()
        logger.info("Alts Loaded")

        storageManager.loadOther()
        logger.info("Other Loaded")

        logger.info("Paragon ${com.paragon.Paragon.Companion.modVersion} Initialised Successfully")
    }

    @EventHandler
    fun postInit(event: FMLPostInitializationEvent) {
        pluginManager.onPostLoad()
    }

    companion object {
        const val modName = "Paragon"
        const val modID = "paragon"
        const val modVersion = "1.0.0"

        @JvmField
        @Mod.Instance
        var INSTANCE = Paragon()
    }

    val eventBus = EventBus()

    // Client stuff
    var logger: Logger = LogManager.getLogger("paragon")
        private set

    var pluginGui: GuiImplementation? = null

    val presenceManager = DiscordPresenceManager()

    // Managers
    lateinit var storageManager: StorageManager
        private set

    lateinit var moduleManager: ModuleManager
        private set

    lateinit var commandManager: CommandManager
        private set

    private lateinit var eventParser: EventFactory

    lateinit var popManager: PopManager
        private set

    lateinit var rotationManager: RotationManager
        private set

    lateinit var friendManager: FriendManager
        private set

    lateinit var altManager: AltManager
        private set

    lateinit var notificationManager: NotificationManager
        private set

    lateinit var capeManager: CapeManager
        private set

    lateinit var pluginManager: PluginManager
        private set

    // GUIs
    lateinit var taskbar: Taskbar
        private set

    lateinit var windows98GUI: Windows98
        private set

    lateinit var configurationGUI: ConfigurationGUI
        private set

    lateinit var console: Console
        private set

    var isParagonMainMenu = false
        set(paragonMainMenu) {
            field = paragonMainMenu
            storageManager.saveOther()
        }

}